src/Nethermind/Nethermind.TxPool/NonceManager.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using Nethermind.Core;
+using Nethermind.Int256;
+
+namespace Nethermind.TxPool;
+
+public class NonceManager : INonceManager
+{
+    private readonly ConcurrentDictionary<Address, AddressNonceManager> _addressNonceManagers = new();
+    private readonly IAccountStateProvider _accounts;
+
+    public NonceManager(IAccountStateProvider accounts)
+    {
+        _accounts = accounts;
+    }
+
+    public UInt256 ReserveNonce(Address address)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        return addressNonceManager.ReserveNonce(_accounts.GetAccount(address).Nonce);
+    }
+
+    public void TxAccepted(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxAccepted();
+        }
+    }
+
+    public void TxRejected(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxRejected();
+        }
+    }
+
+    public void TxWithNonceReceived(Address address, UInt256 nonce)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        addressNonceManager.TxWithNonceReceived(nonce);
+    }
+
+    private class AddressNonceManager
+    {
+        private HashSet<UInt256> _usedNonces = new();
+        private UInt256 _reservedNonce;
+        private UInt256 _currentNonce;
+        private UInt256 _previousAccountNonce;
+        private readonly Mutex _mutex = new();
```

Comment:
`Mutex` is heavy, operation system object. You probably want to use `Monitor`.
But maybe we can changed `_usedNonces` to `ConcurrentHashSet` and don't have lock at all?

Reply:
Without lock we will fail in this scenario(tx A and tx B received in parallel):
1. sendTxA(nonce=null, value=balance+1) => nonce = 0 assigned
2. sendTxB(noce=null, value=balnce-1) => nonce = 1 assigned
3. SubmitTx(A) => invalid
4. SubmitTx(B) => valid
5. tx B is stuck because of nonce gap

Reply:
use `Monitor` then.